docs/contributors/guidelines.rst

```diff
@@ -0,0 +1,205 @@
+
+All of it. All traffic, including HTTP(S), IMAP, SMTP, POP3, should go through
+the front container.
+
+More generally, the front container is responsible for routing that traffic based
+on the incoming port or the HTTP browsing logic. It handles protocol rewriting
+for security, authentication, rejects based on identity or IP address.
+
+How browsing should be managed
+``````````````````````````````
+
+The nginx container is responsible for routing all Web traffic. Web traffic should
+go directly from the nginx container to the target container.
+
+All traffic to a container should be accessed at ``/<container_name>`` for a given
+container. If multiple types of traffic are routed for a specific container, they
+should be accessed at ``/<container_name>/<traffic_type_a>`` and
+``/<container_name>/<traffic_type_b>``, for instance ``/admin/ui`` or
+``/admin/api``.
+
+Traffic directed to ``/`` should be routed to an endpoint on the admin container
+responsible for dynamically redirecting it.
+
+
+How authentication should work
+``````````````````````````````
+
+Authentication should be managed by the nginx container, based on the admin
+container API. The only exception is authentication to the admin container
+itself, that handled directly.
+
+Authentication API in the admin container should contain the logic for
+account management (allowed features, etc.) and rate limiting.
+
+Configuration
+------------
+
+Where should configuration files be stored
+````````````````````````````````````````
+
+Configuration files should ideally be stored in the most standard place for the
+tool being configured. For instance, if the tool usually accepts configuration
+files in ``/etc``, then configuration should be written there.
+
+Should we use default configuration
+``````````````````````````````````
+
+Some tools ship with default configuration, that handles the standard behavior.
+Using this configuration is prone to later changes and unexpected side effects.
+We should always provide all required configuration, including the base files,
+and not rely on default configuration files form the distribution.
+
+For that reason, in case the tool looks for specific files and include them
+automatically, we should overwrite them or delete them.
+
+How should configuration be overridden
+`````````````````````````````````````
+
+Some containers support configuration override. For this feature, we should
+ideally look for conditional configuration inclusion in the configuration syntax
+and use it. If the tool supports multiple methods of overrides, we should use
+the one that supports overriding the more configuration.
+
+In case the tool does not support conditional inclusion, we can add the
+override logic in the `start.py` script.
+
+How much should configuration be documented
+``````````````````````````````````````````
+
+We should not keep default documentation included by the distribution when
+providing configuration files.
```

Comment:
I'm curious about the rationale here. I might have quite some understanding how most parts in Mailu work. But I find it deeply frustrating to go through these bare-bones config files, for instance during bug-hunting. Its a real time waster to use Google in trying to understand what the config is doing.

It might be even worse for new users and potential contributors. Our online documentation is more or less how-to based, with a FAQ based on past incidents and there is small configuration reference. There is no resource telling anybody what most of the internal config means.

This might make the learning curve to steep for users that are experiencing issues or are looking to contribute.

Reply:
I agree we should add more documentation to our config files. But I believe our documentation should state what our specific config does and why we wrote it that way, not include the list of available options and their meaning for every available setting.

Reply:
Yes, you are right.